*: lease timetolive

gyuho opened the pull request:
Fix https://github.com/coreos/etcd/issues/5868.

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -636,6 +643,19 @@ message LeaseGrantResponse {
+  int64 ID = 1;
+}
+
+message LeaseLookupResponse {
```
probably we should also response with all the keys that attached to the lease?

gyuho replied:
Ok will fix.

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -636,6 +643,19 @@ message LeaseGrantResponse {
+  // ID is the lease ID for the granted lease.
+  int64 ID = 2;
+  int64 TTL = 3;
+  string error = 4;
```
we should not put an error here. gRPC can return error at rpc level.

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -206,6 +210,20 @@ func (s *EtcdServer) LeaseGrant(ctx context.Context, r *pb.LeaseGrantRequest) (*
+	if r.ID == int64(lease.NoLease) {
+		return nil, lease.ErrLeaseNotFound
+	}
+	result, err := s.processInternalRaftRequestOnce(ctx, pb.InternalRaftRequest{LeaseLookup: r})
```
why does this go through consensus?

gyuho replied:
Will make it local. Was trying to get expiry from promote lease, but I dropped it, so we don't need to go through raft. Thanks.

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -58,6 +59,9 @@ type Lease interface {
 	Grant(ctx context.Context, ttl int64) (*LeaseGrantResponse, error)
 
+	// Lookup looks up the given lease.
+	Lookup(ctx context.Context, id LeaseID) (*LeaseLookupResponse, error)
```
this needs a better verb since this naming will bubble up to the client-- `client.Lookup()` isn't very "leasey"

gyuho commented:
@xiang90 @heyitsanthony Addressed. PTAL. Thanks.

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -648,6 +648,8 @@ message LeaseRevokeResponse {
 message LeaseKeepAliveRequest {
   // ID is the lease ID for the lease to keep alive.
   int64 ID = 1;
+  // Probe is true, then it only retrieves lease information without renewing.
```
`Retrieve lease information without refreshing the lease.`?

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -656,6 +658,12 @@ message LeaseKeepAliveResponse {
   int64 TTL = 3;
+  // Remaining is the remaining time before lease expiration in seconds.
+  // (only filled in for 'lease probe' operations)
+  int64 remaining = 4;
```
just reuse `TTL`?

heyitsanthony commented on lease/lessor.go:
```diff
@@ -480,6 +481,23 @@ func (l *Lease) refresh(extend time.Duration) {
 func (l *Lease) forever() { l.expiry = forever }
 
+// GetKeys returns all the keys attached to this lease.
+func (l *Lease) GetKeys() []string {
```
`Keys()`

heyitsanthony commented on lease/lessor.go:
```diff
@@ -480,6 +481,23 @@ func (l *Lease) refresh(extend time.Duration) {
+}
+
+// GetRemaining returns all the keys attached to this lease.
+func (l *Lease) GetRemaining() time.Duration {
```
Remaining()

heyitsanthony commented on lease/lessor.go:
```diff
@@ -480,6 +481,23 @@ func (l *Lease) refresh(extend time.Duration) {
+		keys = append(keys, k.Key)
+	}
+	if len(keys) > 1 {
+		sort.Strings(keys)
```
probably shouldn't sort in the server

heyitsanthony commented on lease/lessor.go:
```diff
@@ -480,6 +481,23 @@ func (l *Lease) refresh(extend time.Duration) {
 
+// GetKeys returns all the keys attached to this lease.
+func (l *Lease) GetKeys() []string {
+	var keys []string
```
`keys := make([]string, len(l.itemSet))`?

heyitsanthony commented on lease/leasehttp/http.go:
```diff
@@ -50,64 +50,93 @@ func (h *leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	if err != nil {
-		if err == lease.ErrLeaseNotFound {
-			http.Error(w, err.Error(), http.StatusNotFound)
+	switch lreq.Probe {
```
`if lreq.Probe {} else {}`

heyitsanthony commented on lease/leasehttp/http.go:
```diff
@@ -50,64 +50,93 @@ func (h *leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-		return
+		// TODO: fill out ResponseHeader
+		resp := &pb.LeaseKeepAliveResponse{ID: lreq.ID, TTL: le.TTL, Remaining: int64(le.GetRemaining().Seconds()), Keys: le.GetKeys()}
+		v, err := resp.Marshal()
```
this code down to `w.Write(v)` is common to both paths; should move to after the if/else to deduplicate

heyitsanthony commented on etcdserver/v3_server.go:
```diff
@@ -217,16 +221,59 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
 
+func (s *EtcdServer) LeaseProbe(id lease.LeaseID) (*pb.LeaseKeepAliveResponse, error) {
+	// this will return lease.ErrNotPrimary first, if not primary
+	if _, err := s.lessor.Renew(lease.NoLease); err == lease.ErrLeaseNotFound {
```
why not `s.Leader() == s.ID()`?

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -220,6 +232,55 @@ func (l *lessor) KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAlive
+			return nil, toErr(ctx, err)
+		}
+
+		if nerr := l.newStream(); nerr != nil {
```
why? `probe` always creates a new stream

heyitsanthony commented on etcdctl/ctlv3/command/printer.go:
```diff
@@ -35,6 +35,8 @@ type printer interface {
 	Txn(v3.TxnResponse)
 	Watch(v3.WatchResponse)
 
+	LeaseProbe(v3.LeaseProbeResponse)
```
`Probe`

heyitsanthony commented on etcdserver/api/v3rpc/lease.go:
```diff
@@ -73,6 +73,17 @@ func (ls *LeaseServer) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) erro
 		resp := &pb.LeaseKeepAliveResponse{ID: req.ID, Header: &pb.ResponseHeader{}}
 		ls.hdr.fill(resp.Header)
 
+		if req.Probe {
```
have methods `LeaseServer.probe(id) (*LeaseKeepAliveResponse, error)` and `LeaseServer.renew(id) (*LeaseKeepAliveReponse, error)` instead of handling both directly in this function?

heyitsanthony commented on etcdserver/api/v3rpc/lease.go:
```diff
@@ -73,6 +73,17 @@ func (ls *LeaseServer) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) erro
+			resp.TTL = lresp.TTL
+			resp.Remaining = lresp.Remaining
+			resp.Keys = lresp.Keys
+			return stream.Send(resp)
```
don't bail out of the stream, just keep accepting requests like renew? otherwise this seems like it would leak

heyitsanthony commented on proxy/grpcproxy/lease.go:
```diff
@@ -66,6 +66,9 @@ func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) error
 			if err != nil {
 				return
 			}
+			if rr.Probe {
```
probably not necessary; no reason to stop processing requests on the stream

heyitsanthony commented on proxy/grpcproxy/lease.go:
```diff
@@ -78,5 +81,8 @@ func (lp *leaseProxy) LeaseKeepAlive(stream pb.Lease_LeaseKeepAliveServer) error
 		if err != nil {
 			return err
 		}
+		if rr.Remaining > 0 { // probe response
```
same here; keep processing requests instead of bailing out

heyitsanthony commented on etcdctl/ctlv3/command/lease_command.go:
```diff
@@ -134,3 +135,33 @@ func leaseKeepAliveCommandFunc(cmd *cobra.Command, args []string) {
+	id, err := strconv.ParseInt(args[0], 16, 64)
+	if err != nil {
+		ExitWithError(ExitBadArgs, fmt.Errorf("bad lease ID arg (%v), expecting ID in Hex", err))
+	}
```
this path happens several times in `lease_command.go`; have something like `leaseFromArgs(args []string) v3.LeaseID` that handles all the error checking / exiting for probe/renew/revoke?

xiang90 commented:
@gyuho 

Discussed with @heyitsanthony. The current approach does not work well since we will renew the lease when probing.

I think we should still have a separate user facing RPC for getting the information of a lease. Probably we should call it TimeToLive

```proto
message LeaseTimeToLiveRequest {
  // ID is the lease ID for the lease to keep alive.
  int64 ID = 1;
}

message LeaseTimeToLiveResponse {
  ResponseHeader header = 1;
  // ID is the lease ID from the keep alive request.
  int64 ID = 2;
  // TTL is the remaining time-to-live for the lease.
  int64 TTL = 3;
  // Keys is the list of keys attached to this lease.
  repeated string keys = 5;
}
```

gyuho commented:
@xiang90 Ok will try that approach today. Thanks!

gyuho commented:
Now probe is a separate RPC but uses the same logic as keep-alive.
PTAL. Thanks.

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -104,8 +104,14 @@ service Lease {
     };
   }
 
+  rpc LeaseProbe(LeaseProbeRequest) returns (LeaseProbeResponse) {
```
LeaseTimeToLive

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
   int64 TTL = 3;
 }
 
+message LeaseProbeRequest {
```
LeaseTimeToLiveRequest

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+  int64 ID = 1;
+}
+
+message LeaseProbeResponse {
```
LeaseTimeToLiveResponse

xiang90 commented on etcdserver/v3_server.go:
```diff
@@ -235,21 +295,13 @@ func (s *EtcdServer) LeaseRenew(id lease.LeaseID) (int64, error) {
 	}
-
-	for _, url := range leader.PeerURLs {
-		lurl := url + "/leases"
```
define lease as a const? now we have two places use it.

gyuho replied:
Yeah will define that in `v2http`.

xiang90 replied:
this is not v2 stuff. should be defined in our lease http pkg.

xiang90 commented on lease/leasehttp/http.go:
```diff
@@ -44,28 +44,59 @@ func (h *leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-		if err == lease.ErrLeaseNotFound {
-			http.Error(w, err.Error(), http.StatusNotFound)
+	case "/leases/internal":
+		lreq := pb.LeaseProbeRequest{}
```
let's not use user facing proto here. let's define an internal lease proto for this. then we can easily extend it.

gyuho replied:
Could you explain `define an internal lease proto`?

Then we want to define `LeaseInternalRequest` rather than `LeaseTimeToLiveRequest`?

xiang90 replied:
LeaseTimeToLiveRequest is the user facing request type. It should be the contract between etcd server and user. We should not use that proto between etcd servers. Now we use it between etcd follower and server communication. So we need to change it. The reason is that user facing proto is fixed once defined. We can always evolve the internal one as we want. It is good to separate them out.

xiang90 commented:
LGTM in general. Some minor issues need to be fixed.

gyuho commented:
@xiang90 All addressed. PTAL. Thanks!

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,28 @@ message LeaseKeepAliveResponse {
   int64 TTL = 3;
 }
 
+message LeaseInternalRequest {
```
move this to leasepb?

xiang90 commented:
rename all probe to timetolive?

gyuho commented:
@xiang90 Just renamed them all. PTAL. Thanks!

xiang90 commented on clientv3/lease.go:
```diff
@@ -44,6 +44,15 @@ type LeaseKeepAliveResponse struct {
 	TTL int64
 }
 
+// LeaseTimeToLiveResponse is used to convert the protobuf lease probe response.
```
probe -> keepalive

xiang90 commented on etcdserver/v3_server.go:
```diff
@@ -219,14 +223,70 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
+
+func (s *EtcdServer) LeaseTimeToLive(ctx context.Context, r *pb.LeaseTimeToLiveRequest) (*pb.LeaseTimeToLiveResponse, error) {
+	if s.Leader() == s.ID() {
+		// primary; probe directly from leader
```
probe -> keepalive

xiang90 commented on etcdserver/v3_server.go:
```diff
@@ -219,14 +223,70 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
+
+	var lresp *pb.LeaseTimeToLiveResponse
+	donec := make(chan struct{})
+	go func() {
```
why do we need this go routine?

gyuho replied:
Otherwise, we never use the given context? Should we just not use that context?

xiang90 replied:
we probably should make leasehttp.TimeToLive respect the context.

xiang90 commented on lease/leasehttp/http.go:
```diff
@@ -44,28 +50,59 @@ func (h *leaseHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+			return
+		}
+		// TODO: fill out ResponseHeader
+		resp := &pb.LeaseTimeToLiveResponse{
```
the response should be internalResponse which contains leaseTimeToLiveResponse

gyuho replied:
So define `leasehttp.LeaseInternalResponse`?

xiang90 replied:
yes.

xiang90 commented:
There are still a few probes in the code. let's fix them up.

xiang90 commented on lease/leasehttp/http.go:
```diff
@@ -111,3 +148,42 @@ func RenewHTTP(id lease.LeaseID, url string, rt http.RoundTripper, timeout time.
 }
+
+// TimeToLiveHTTP probes the given lease ID.
+func TimeToLiveHTTP(id lease.LeaseID, url string, rt http.RoundTripper, timeout time.Duration) (*pb.LeaseTimeToLiveResponse, error) {
```
accept context in this func? remove timeout. use context.Deadline to set the current timeout

if the context times out, cancel http request

gyuho commented:
@xiang90 PTAL.

xiang90 commented on etcdserver/v3_server.go:
```diff
@@ -219,14 +224,63 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
+		var iresp *leasepb.LeaseInternalResponse
+		iresp, err = leasehttp.TimeToLiveHTTP(ctx, lease.LeaseID(r.ID), lurl, s.peerRt)
+		if err == nil {
+			if iresp.LeaseTimeToLiveResponse.Header == nil {
```
why do we need to do this. this seems to be wrong.

xiang90 commented on lease/leasehttp/http.go:
```diff
@@ -111,3 +152,61 @@ func RenewHTTP(id lease.LeaseID, url string, rt http.RoundTripper, timeout time.
+	}
+	req.Header.Set("Content-Type", "application/protobuf")
+
+	cancel := httputil.RequestCanceler(req)
```
```go
go func() {
    // liste on context events
    // call cancel if context is done
}

```

xiang90 replied:
then we can simplify code from 172 to 188. we need to do this since we should respect context for the reading at line 198.

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+  ResponseHeader header = 1;
+  // ID is the lease ID from the keep alive request.
+  int64 ID = 2;
+  // TTL is the new time-to-live for the lease.
```
TTL is the remaining TTL for the lease

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+  // TTL is the new time-to-live for the lease.
+  int64 TTL = 3;
+  // Remaining is the remaining time in seconds before expiration.
+  int64 remaining = 4;
```
remove remaining

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+  int64 TTL = 3;
+  // Remaining is the remaining time in seconds before expiration.
+  int64 remaining = 4;
+  // Keys is the list of keys attached to this lease.
```
we can add the grantedTTL if we want.

gyuho replied:
Ok will remove `remaining`. I think `remaining` can be useful with longer TTL and user wants to know what's the remainder. Ok we can easily add later.

xiang90 replied:
We need to return the remaining by default. TTL = time to live. Not the granted one. We can return the granted one as grantedTTL.

Basically remaining - > ttl; ttl-> grantedTTL in your existing code.

xiang90 commented:
lgtm in general. defer to @heyitsanthony. I have to prepare for my flight. :(

gyuho commented:
@xiang90 Have a great trip. I will address your feedback by tonight. Thanks.

gyuho commented:
PTAL. I put the `ReadAll` lines into goroutine. I think checking context inside goroutine makes it complicated as well, because we still need to receive the error from the goroutine.

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+message LeaseTimeToLiveRequest {
+  // ID is the lease ID for the lease to keep alive.
+  int64 ID = 1;
+}
```
`bool keys` option so it won't send all the attached keys by default?

gyuho replied:
Ok will add.

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +664,23 @@ message LeaseKeepAliveResponse {
+  int64 ID = 2;
+  // TTL is the remaining TTL for the lease.
+  int64 TTL = 3;
+  // GrantedTTL is the initial granted time in seconds upon lease creation.
```
`creation/renewal`?

heyitsanthony commented on etcdctl/README.md:
```diff
@@ -344,6 +344,32 @@ lease 32695410dcc0ca0 keepalived with TTL(100)
+lease 5ee156e4d3c12c08 granted with TTL(500s), remaining(485s), keys([foo1 foo2])
+
+./etcdctl lease timetolive 5ee156e4d3c12c08 --write-out=json
+{"cluster_id":10106783228653711634,"member_id":15241412574772223713,"revision":3,"raft_term":2,"id":6836841250130308104,"ttl":500,"remaining":483,"keys":["foo1","foo2"]}
```
json isn't base64 encoding the keys?

gyuho replied:
`golang.org` says it's base64-encoded only when defined as `[]byte`.

xiang90 replied:
we should define keys as []byte then :)

gyuho replied:
Ok will try it. Thanks.

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -220,6 +232,33 @@ func (l *lessor) KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAlive
+		if isHaltErr(cctx, err) {
+			return nil, toErr(ctx, err)
+		}
+		if nerr := l.newStream(); nerr != nil {
```
this rpc shouldn't interact with the keepalive stream

gyuho commented:
@heyitsanthony @xiang90 All addressed. PTAL. Thanks.

xiang90 commented on clientv3/lease.go:
```diff
@@ -170,6 +182,36 @@ func (l *lessor) Revoke(ctx context.Context, id LeaseID) (*LeaseRevokeResponse,
+				TTL:            resp.TTL,
+				GrantedTTL:     resp.GrantedTTL,
+			}
+			if len(resp.Keys) > 0 {
```
we can remove this if statement. range over a zero length slice is OK.

xiang90 commented on etcdctl/ctlv3/command/printer.go:
```diff
@@ -159,6 +161,14 @@ func (s *simplePrinter) Watch(resp v3.WatchResponse) {
+	for i := range resp.Keys {
+		ks[i] = string(resp.Keys[i])
+	}
+	fmt.Printf("lease %016x granted with TTL(%ds), remaining(%ds), keys(%v)\n", resp.ID, resp.GrantedTTL, resp.TTL, ks)
```
keys -> attached keys

xiang90 commented:
Thanks @gyuho!

LGTM. Defer to @heyitsanthony

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +665,25 @@ message LeaseKeepAliveResponse {
+  // GrantedTTL is the initial granted time in seconds upon lease creation/renewal.
+  int64 grantedTTL = 4;
+  // Keys is the list of keys attached to this lease.
+  repeated string keys = 5;
```
this has to be `repeated bytes` since "A string must always contain UTF-8 encoded or 7-bit ASCII text." but etcd keys can be arbitrary binary strings

heyitsanthony commented on etcdctl/ctlv3/command/printer.go:
```diff
@@ -159,6 +161,14 @@ func (s *simplePrinter) Watch(resp v3.WatchResponse) {
+	for i := range resp.Keys {
+		ks[i] = string(resp.Keys[i])
+	}
+	fmt.Printf("lease %016x granted with TTL(%ds), remaining(%ds), attached keys(%v)\n", resp.ID, resp.GrantedTTL, resp.TTL, ks)
```
if len(resp.keys) == 0, only print granted and remaining? it's sort of misleading to do --keys=false and then it says attached keys == []

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -170,6 +182,34 @@ func (l *lessor) Revoke(ctx context.Context, id LeaseID) (*LeaseRevokeResponse,
+			return gresp, nil
+		}
+		if isHaltErr(cctx, err) {
+			return nil, toErr(ctx, err)
```
`toErr(cctx, err)`

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -68,6 +77,9 @@ type Lease interface {
 	KeepAliveOnce(ctx context.Context, id LeaseID) (*LeaseKeepAliveResponse, error)
 
+	// TimeToLive retrieves the lease information of the given lease ID.
+	TimeToLive(ctx context.Context, id LeaseID, keys bool) (*LeaseTimeToLiveResponse, error)
```
`TimeToLive(ctx context.Context, id LeaseID, opts ...LeaseOption) (*LeaseTimeToLiveResponse, error)`?

heyitsanthony commented on clientv3/integration/lease_test.go:
```diff
@@ -455,3 +457,48 @@ func TestLeaseKeepAliveTTLTimeout(t *testing.T) {
+	if lresp.GrantedTTL != int64(10) {
+		t.Fatalf("GrantedTTL expected %d, got %d", 10, lresp.GrantedTTL)
+	}
+	if lresp.TTL == 0 || lresp.TTL > lresp.GrantedTTL {
```
will `Remaining.Seconds()` round down to 0 when cast to `int64` for `TTL` if the lease has less than a second left?

xiang90 replied:
I think round down to 0 is OK. It means the lease will be removed soon enough. We only give users second level accuracy. And TTL is the lower bound.

xiang90 replied:
I think round down to 0 is OK. It means the lease will be removed soon enough. We only give users second level accuracy. And TTL is the upper bound.

heyitsanthony replied:
So long as TTL == 0 is treated as [0, 1) seconds left instead of the key being expired, it's fine. My main complaint is it's being treated as an error here.

gyuho commented:
@heyitsanthony All addressed. PTAL.

1. changed `Keys` to `[][]byte`
2. changed 'TTL` to `float64` to not truncate milliseconds
3. changed output function to print out `attached keys` only when `--keys` flag is passed
4. added `LeaseOption` to clientv3
5. fixed `toErr`

Thanks!

gyuho commented on scripts/genproto.sh:
```diff
@@ -60,6 +60,7 @@ for dir in ${DIRS}; do
 		protoc --gofast_out=plugins=grpc,import_prefix=github.com/coreos/:. -I=.:"${GOGOPROTO_PATH}":"${COREOS_ROOT}":"${GRPC_GATEWAY_ROOT}/third_party/googleapis" *.proto
 		sed -i.bak -E "s/github\.com\/coreos\/(gogoproto|github\.com|golang\.org|google\.golang\.org)/\1/g" *.pb.go
 		sed -i.bak -E 's/github\.com\/coreos\/(errors|fmt|io)/\1/g' *.pb.go
+		sed -i.bak -E 's/import math \"github\.com\/coreos\/math\"//g' *.pb.go
```
This is needed when proto file has `double` field, the generated Go code inserts `import math "github.com/coreos/math"` line.

xiang90 commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +665,25 @@ message LeaseKeepAliveResponse {
+  // ID is the lease ID from the keep alive request.
+  int64 ID = 2;
+  // TTL is the remaining TTL in seconds for the lease.
+  double TTL = 3;
```
can be change this back to int64? what is the reason to make this double? this change is inconsistent with other ttl fields.

gyuho replied:
@xiang90  One concern was that TTL of <1second was being rounded down to 0 second with type-cast.
I guess we can just use `int64` to be consistent?

xiang90 replied:
round down to 0 is fine. We need to clearly doc about it.

gyuho replied:
sure, will use `int64` with more documentation around it.

gyuho commented:
Now `TTL` is `int64`

heyitsanthony commented on clientv3/lease.go:
```diff
@@ -44,6 +44,22 @@ type LeaseKeepAliveResponse struct {
+	ID LeaseID `json:"id"`
+
+	// TTL is the remaining TTL in seconds for the lease.
+	// Milliseconds will be rounded off.
```
`TTL is the remaining TTL in seconds for the lease; the lease will expire in under TTL+1 seconds.`?

heyitsanthony commented on etcdserver/etcdserverpb/rpc.proto:
```diff
@@ -658,6 +665,26 @@ message LeaseKeepAliveResponse {
+  int64 ID = 2;
+  // TTL is the remaining TTL in seconds for the lease.
+  // Milliseconds will be rounded off.
+  int64 TTL = 3;
```
`TTL is the remaining TTL in seconds for the lease; the lease will expire in under TTL+1 seconds.`?

xiang90 replied:
@heyitsanthony i belive yours is better. we do not really expose ms accuracy anyway. we want users to understand a 0 TTL does not mean the lease is expired.

heyitsanthony commented:
lgtm after fixing minor doc nit. Thanks!

gyuho commented:
Just updated the comments. Will merge after greens. Thanks a lot!
